### PR TITLE
🐛 OPRUN-4217: OwnNamespace default handling

### PR DESCRIPTION
**Pull request by perdasilva**
# Description

Currently, if a bundle only supports OwnNamespace the `watchNamespace` configuration will no be present for that bundle and the bundle will be automatically installed in OwnNamespace mode (by setting the targetNamespaces to the install namespace). While this is intuitive and makes life easier for the user, it can can lead to a break in default behavior if the package decided to add AllNamespaces support. For instance:

v1 only supports OwnNamespace install mode. The ClusterExtension has no configuration so the bundle is installed in OwnNamespace mode.

v1.1 adds AllNamespaces support and the ClusterExtension detects the upgrade and initiates it. The default is now AllNamespaces mode. So, the operator is now installed in AllNamespace mode without knowledge or acceptance by the user.

To address this issue, we modify the defaulting and configuration behavior for this case by requiring `watchNamespace` to be set and removing the defaulting behavior, i.e. making it undefined. I.e. only bundles with AllNamespaces mode can be installed without `watchNamespace` configuration. This means that when v2 comes out with AllNamespaces support, the configuration will already be present in the ClusterExtension and v2 will be installed in the same mode. The user can then remove the configuration if they wish to change install modes.

So, in this PR, we:
- Add some additional unit tests around watchNamespaces being null/empty (this is not exactly related but was identified while writing this PR)
- Modify the configuration validation to ensure that watchNamespaces is required when OwnNamespace is the only supported install mode (we include MultiNamespaces in the tests for completeness but this install mode will never be supported)
- Modify the defaulting behavior in the reg+v1 renderer to ensure no defaults are set for bundles that don't support AllNamespaces install mode
- Add `ExtensionConfigBytes()` method to `ClusterExtension` to easily grab user configuration
- Update `UnmarshalConfig` to treat `nil` as `{}` to force schema validation (catch missing required values, etc.)
- Update `RegistryV1ManifestProvider` tests to ensure bundle configuration is being properly validated
- Add single-namespace- and own-namespace-operator fixtures and update e2e tests to check for bad configuration as well as good

## Reviewer Checklist

- [ ] API Go Documentation
- [ ] Tests: Unit Tests (and E2E Tests, if appropriate)
- [ ] Comprehensive Commit Messages
- [ ] Links to related GitHub Issue(s)

### CODE REVIEW

File: testdata/images/bundles/single-namespace-operator/v1.0.0/manifests/singlenamespaceoperator.clusterserviceversion.yaml [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+    createdAt: "2024-10-24T19:21:40Z"
+    operators.operatorframework.io/builder: operator-sdk-v1.34.1
+    operators.operatorframework.io/project_layout: go.kubebuilder.io/v4
+  name: singlenamespaceoperator.v1.0.0
+  namespace: placeholder
+spec:
+  apiservicedefinitions: {}
+  customresourcedefinitions:
+  owned:
+    - description: Configures subsections of Alertmanager configuration specific to each namespace
```
[the remark this answers is not in the document]
Comment (perdasilva) [reply testdata/images/bundles/single-namespace-operator/v1.0.0/manifests/singlenamespaceoperator.clusterserviceversion.yaml]: I've updated it to something else

File: testdata/images/bundles/own-namespace-operator/v1.0.0/manifests/ownnamespaceoperator.clusterserviceversion.yaml [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+              containers:
+              - name: busybox
+                image: busybox:1.36
+                command:
+                - 'sleep'
+                - '1000'
+                securityContext:
+                  runAsUser: 1000
+                  runAsNonRoot: true
+                serviceAccountName: simple-bundle-manager
```
[the remark this answers is not in the document]
Comment (perdasilva) [reply testdata/images/bundles/own-namespace-operator/v1.0.0/manifests/ownnamespaceoperator.clusterserviceversion.yaml:68]: fixed

File: internal/operator-controller/rukpak/bundle/config.go [hunk 3]
```diff
@@ -22,10 +22,10 @@ type Config struct {
 // - rv is nil
 // - bytes is not a valid YAML/JSON object
 // - bytes is a valid YAML/JSON object but does not follow the registry+v1 schema
-// if bytes is nil a nil bundle.Config is returned
+// if bytes is nil it will be treated as an empty json object ({})
 func UnmarshallConfig(bytes []byte, rv1 RegistryV1, installNamespace string) (*Config, error) {
```
[the remark this answers is not in the document]
Comment (perdasilva) [reply internal/operator-controller/rukpak/bundle/config.go]: Fixed

File: internal/operator-controller/rukpak/bundle/config.go [hunk 4]
```diff
@@ -83,20 +83,19 @@ func validateConfig(config *Config, installNamespace string, bundleInstallModeSe
...
-// AllNamespaces install mode is not supported and SingleNamespace is supported
+// AllNamespaces install mode is not supported and SingleNamespace and/or OwnNamespace is supported
 func isWatchNamespaceConfigRequired(bundleInstallModeSet sets.Set[v1alpha1.InstallMode]) bool {
-	return !bundleInstallModeSet.Has(v1alpha1.InstallMode{Type: v1alpha1.InstallModeTypeAllNamespaces, Supported: true}) &&
-		bundleInstallModeSet.Has(v1alpha1.InstallMode{Type: v1alpha1.InstallModeTypeSingleNamespace, Supported: true})
+	return isWatchNamespaceConfigSupported(bundleInstallModeSet) &&
+		!bundleInstallModeSet.Has(v1alpha1.InstallMode{Type: v1alpha1.InstallModeTypeAllNamespaces, Supported: true})
 }
 
 // formatUnmarshallError format JSON unmarshal errors to be more readable
```
[the remark this answers is not in the document]
Comment (perdasilva) [reply internal/operator-controller/rukpak/bundle/config.go]: Fixed

File: internal/operator-controller/rukpak/render/render.go [hunk 5]
```diff
@@ -160,10 +160,10 @@ func validateTargetNamespaces(rv1 *bundle.RegistryV1, installNamespace string, t
 		// in case only the MultiNamespace install mode is supported by the bundle.
 		// If AllNamespaces mode is supported, the default will be [""] -> watch all namespaces
 		// If only OwnNamespace is supported, the default will be [install-namespace] -> only watch the install/own namespace
```
Comment (camilamacedo86) [inline internal/operator-controller/rukpak/render/render.go:162]:
```suggestion
		// If AllNamespaces is not supported, no default is applied and configuration must provide the target namespace(s)
```

File: api/v1/clusterextension_types.go [hunk 6]
```diff
@@ -535,6 +535,23 @@ type ClusterExtension struct {
 	Status ClusterExtensionStatus `json:"status,omitempty"`
```
Comment (camilamacedo86) [inline api/v1/clusterextension_types.go:535]: I think we need to change here: https://github.com/operator-framework/operator-controller/blob/926d57e1e6559f7ac0ef69a47b3d50d96975ec38/docs/draft/howto/single-ownnamespace-install.md?plain=1#L61

With somthing like

`- *AllNamespaces*: `watchNamespace` is empty, or not set **and** the bundle supports the AllNamespaces install mode. Bundles that lack AllNamespaces support now require explicit configuration.`

File: internal/operator-controller/applier/provider_test.go [hunk 7]
```diff
@@ -330,11 +357,53 @@ func Test_RegistryV1ManifestProvider_SingleOwnNamespaceSupport(t *testing.T) {
...
+				Config: &ocv1.ClusterExtensionConfig{
+					ConfigType: ocv1.ClusterExtensionConfigTypeInline,
+					Inline: &apiextensionsv1.JSON{
+						Raw: []byte(`{"watchNamespace": "install-namespace"}`),
+					},
+				},
 			},
 		})
 		require.NoError(t, err)
 	})
```
Comment (camilamacedo86) [inline internal/operator-controller/applier/provider_test.go:361]:
```suggestion
	// Checks that OwnNamespace bundles use the install namespace as targetNamespaces
	// when watchNamespace is the same. (targetNamespaces == [install-namespace] and no error)
	t.Run("sets targetNamespaces to the install namespace for {OwnNamespace} bundles", func(t *testing.T) {
		expectedTargetNamespace := "install-namespace"
		provider := applier.RegistryV1ManifestProvider{
			BundleRenderer: render.BundleRenderer{
				ResourceGenerators: []render.ResourceGenerator{
					func(rv1 *bundle.RegistryV1, opts render.Options) ([]client.Object, error) {
						require.Equal(t, []string{expectedTargetNamespace}, opts.TargetNamespaces)
						return nil, nil
					},
				},
			},
			IsSingleOwnNamespaceEnabled: true,
		}
		bundleFS := bundlefs.Builder().WithPackageName("test").
			WithCSV(clusterserviceversion.Builder().WithInstallModeSupportFor(v1alpha1.InstallModeTypeOwnNamespace).Build()).Build()

		_, err := provider.Get(bundleFS, &ocv1.ClusterExtension{
			Spec: ocv1.ClusterExtensionSpec{
				Namespace: expectedTargetNamespace,
				Config: &ocv1.ClusterExtensionConfig{
					ConfigType: ocv1.ClusterExtensionConfigTypeInline,
					Inline: &apiextensionsv1.JSON{
						Raw: []byte(`{"watchNamespace": "install-namespace"}`),
					},
				},
			},
		})
		require.NoError(t, err)
	})
```

WDYT?

File: api/v1/clusterextension_types.go [hunk 8]
```diff
@@ -535,6 +535,23 @@ type ClusterExtension struct {
...
+	case ClusterExtensionConfigTypeInline:
+		if e.Spec.Config.Inline != nil {
+			return e.Spec.Config.Inline.Raw
+		}
+		return nil
+	default:
+		return nil
+	}
+}
+
```
Comment (tmshort) [inline api/v1/clusterextension_types.go]: Is this the right place for this? A function within a `_types.go` file?

Comment (perdasilva) [reply api/v1/clusterextension_types.go]: I'm not sure. I think the method makes sense. But I have no strong opinions on whether it should be factored out into its own file or not, or what that file should be called.

File: internal/operator-controller/applier/provider_test.go (hunk 7 above)
Comment (perdasilva) [reply internal/operator-controller/applier/provider_test.go:361]: If I got it right, what was missing was checking that the appropriate option was being set (i.e. `require.Equal(t, []string{expectedTargetNamespace}, opts.TargetNamespaces)`), right?

I've updated the `accepts bundles with {OwnNamespace} install modes when the appropriate configuration is given` to cover that

File: internal/operator-controller/applier/provider_test.go [hunk 9]
```diff
@@ -321,20 +312,89 @@ func Test_RegistryV1ManifestProvider_SingleOwnNamespaceSupport(t *testing.T) {
...
-			WithCSV(clusterserviceversion.Builder().WithInstallModeSupportFor(v1alpha1.InstallModeTypeOwnNamespace).Build()).Build()
+			WithCSV(clusterserviceversion.Builder().WithInstallModeSupportFor(v1alpha1.InstallModeTypeSingleNamespace).Build()).Build()
+
 		_, err := provider.Get(bundleFS, &ocv1.ClusterExtension{
 			Spec: ocv1.ClusterExtensionSpec{
 				Namespace: "install-namespace",
 			},
 		})
+		require.Error(t, err)
+		require.Contains(t, err.Error(), "required field \"watchNamespace\" is missing")
```
Comment (camilamacedo86) [inline internal/operator-controller/applier/provider_test.go:329]:
``` suggestion
		require.Contains(t, err.Error(), "watchNamespace is required for SingleNamespace install mode")
```

What about adding an install mode and having unique errors, not only to make it easier to understand but also to make troubleshooting easier?

File: internal/operator-controller/applier/provider_test.go [hunk 10]
```diff
@@ -321,20 +312,89 @@ func Test_RegistryV1ManifestProvider_SingleOwnNamespaceSupport(t *testing.T) {
...
+		}
+		bundleFS := bundlefs.Builder().WithPackageName("test").
+			WithCSV(clusterserviceversion.Builder().WithInstallModeSupportFor(v1alpha1.InstallModeTypeOwnNamespace).Build()).Build()
+		_, err := provider.Get(bundleFS, &ocv1.ClusterExtension{
+			Spec: ocv1.ClusterExtensionSpec{
+				Namespace: "install-namespace",
+			},
+		})
+		require.Error(t, err)
+		require.Contains(t, err.Error(), "required field \"watchNamespace\" is missing")
```
Comment (camilamacedo86) [inline internal/operator-controller/applier/provider_test.go:374]:
```suggestion
		require.Contains(t, err.Error(), "watchNamespace is required for OwnNamespace install mode")
```
What about adding an install mode and having unique errors, not only to make it easier to understand but also to make troubleshooting easier?

File: internal/operator-controller/rukpak/bundle/config_test.go [hunk 11]
```diff
@@ -91,16 +91,28 @@ func Test_UnmarshallConfig(t *testing.T) {
...
-			rawConfig:             []byte(`{"watchNamespace": "some-namespace"}`),
-			expectedErrMessage:    "unknown field \"watchNamespace\"",
+			rawConfig:             []byte(`{"watchNamespace": null}`),
+			expectedErrMessage:    "required field \"watchNamespace\" is missing",
+		},
+		{
+			name:                  "reject with required field when install modes {OwnNamespace} and watchNamespace is missing",
+			supportedInstallModes: []v1alpha1.InstallModeType{v1alpha1.InstallModeTypeOwnNamespace},
+			rawConfig:             []byte(`{}`),
+			expectedErrMessage:    "required field \"watchNamespace\" is missing",
```
Comment (camilamacedo86) [inline internal/operator-controller/rukpak/bundle/config_test.go:103]:
```suggestion
			expectedErrMessage:    "watchNamespace is required for OwnNamespace install mode",
```

What about adding an install mode and having unique errors, not only to make it easier to understand but also to make troubleshooting easier?

File: internal/operator-controller/rukpak/bundle/config_test.go [hunk 12]
```diff
@@ -91,16 +91,28 @@ func Test_UnmarshallConfig(t *testing.T) {
...
 		},
 		{
-			name:                  "reject with unknown field when install modes {MultiNamespace, OwnNamespace}",
+			name:                  "reject with required field when install modes {MultiNamespace, OwnNamespace} and watchNamespace is null",
 			supportedInstallModes: []v1alpha1.InstallModeType{v1alpha1.InstallModeTypeMultiNamespace, v1alpha1.InstallModeTypeOwnNamespace},
-			rawConfig:             []byte(`{"watchNamespace": "some-namespace"}`),
-			expectedErrMessage:    "unknown field \"watchNamespace\"",
+			rawConfig:             []byte(`{"watchNamespace": null}`),
+			expectedErrMessage:    "required field \"watchNamespace\" is missing",
```
Comment (camilamacedo86) [inline internal/operator-controller/rukpak/bundle/config_test.go]:
```suggestion
			expectedErrMessage:    "watchNamespace is required for OwnNamespace install mode",
```

File: api/v1/clusterextension_types.go (hunk 6 above)
Comment (perdasilva) [reply api/v1/clusterextension_types.go:535]: I've done a more comprehensive doc update ^^ thanks for pointing this out - I totally forgot about it

File: internal/operator-controller/applier/provider_test.go (hunk 9 above)
Comment (perdasilva) [reply internal/operator-controller/applier/provider_test.go:329]: We don't want to surface install mode concerns through the errors. Treat it as configuration.

File: internal/operator-controller/applier/provider_test.go (hunk 10 above)
Comment (perdasilva) [reply internal/operator-controller/applier/provider_test.go:374]: I don't think we want to surface these concerns up. It's not that bundle supports this or that. It's configuration is required/optional, the value is valid/invalid, etc. Install modes are a v0 thing. We don't want to leak it here I don't think.

File: api/v1/clusterextension_types.go (hunk 8 above)
Comment (camilamacedo86) [reply api/v1/clusterextension_types.go]: If we do an analogy with the OO (object-oriented) way, it would live inside the class itself;  
i.e. something like `MyObject.GetBytes()`. So yeah, thinking in DDD/OO style, it seems to fit here.

But I have one concern. Since this is in the API package, it becomes public,  
and once it’s public, we can’t easily change or remove it later.

So unless we’re 100% sure we want to keep and support this method as part of the public API,  
it might be better to move it to an internal package — or avoid exporting it for now.

**Review by camilamacedo86 — COMMENTED** [review]
Great work and amazing catch 🎉  

I have only one real concern → https://github.com/operator-framework/operator-controller/pull/2283/files#r2456773553  
I think we should move that part out of the public API.  

The other things are just small nits — it would be nice if we could include the install modes in the error messages, so each case has a more unique and clearer message. That usually helps a lot when troubleshooting.  

Beyond that, it looks like a very solid solution to me 🚀

File: api/v1/clusterextension_types.go (hunk 8 above)
Comment (joelanford) [reply api/v1/clusterextension_types.go]: I'm trying to remember the specifics, but I remember it being inadvisable to attach functionality directly to the API types in the API package.

I _think_ it was mostly related to introducing coupling with shared dependencies that makes it hard/impossible for other projects to import the API.

Should we stick to type definitions in the API package and define a separate package that has helpers like this?

Comment (joelanford) [reply api/v1/clusterextension_types.go]: Also, this function doesn't seem particularly well-suited to handle potential future extensions of sourcing configs from non-inline sources. Sourcing configs from secrets, for example, would require a kubernetes secret client, which we cannot attach to the type as a field and which we are not able to pass into the current function as a parameter.

Comment (camilamacedo86) [reply api/v1/clusterextension_types.go]: >Should we stick to type definitions in the API package and define a separate package that has helpers like this?

@joelanford  
Will those helpers become public as well?  

- If yes, then maybe it makes sense to keep them in the API file.  
- If not, then that actually addresses my concern above. 🙂

Comment (joelanford) [reply api/v1/clusterextension_types.go]: fwiw, this is what I had come up with as a helper in my prototype: https://github.com/joelanford/operator-controller/blob/5f255eef69b424e09d2630366059944ba5875411/internal/values/values.go

Comment (perdasilva) [reply api/v1/clusterextension_types.go]: I'll remove it

**Comment by perdasilva** [comment]
> Just because of the new method added to the public API (exported). I think we should move it out of the API package. → https://github.com/operator-framework/operator-controller/pull/2283/files#r2456565041

I've removed that and made it a private utility closer to where it is used
